crates/ide/src/hover/tests.rs

```diff
@@ -8602,3 +8602,31 @@ fn test() {
         "#]],
     );
 }
+
+#[test]
+fn issue_17871() {
+    check(
+        r#"
+trait T {
+    fn f<A>();
+}
+
+struct S {}
+impl T for S {
+    fn f<A>() {}
+}
+
+fn main() {
+    let x$0 = S::f::<i32>;
+}
+"#,
+        expect![[r#"
+            *x*
+
+            ```rust
+            // size = 0, align = 1
+            let x: fn f<S, i32>()
```

Comment:
Hmm, this does look like the wrong way to render this doesn't it. Ideally we'd render this as `fn <S as T>::f::<i32>()` or so?

Reply:
Yes, I think so, too. Currently we are rendering such container type into `<>` when rendering associated function

rust-analyzer 1.80.0 (05147895 2024-07-21)
![image](https://github.com/user-attachments/assets/7049e36c-2c06-4c86-ad4b-97448794bff5)

And we can find such things in many places of our test cases like
https://github.com/rust-lang/rust-analyzer/blob/2b866390189d99191c51e1fe154e94a933c24b0f/crates/hir-ty/src/tests/traits.rs#L1108

I tried to fix them altogather when I was dealing with this issue, but it seems that our generic displaying has many problems that should be dealt with - especially when there are 'parent' generic parameters -, so I just turned to fix this panic with minimal touch and file the issue in proper way 🤔

Reply:
Yes, there was no need to do all of that in this PR given its separate issues

Reply:
One thing I think that strange is;

We are rendering generics in many cases with `fn hir_fmt_generic_arguments`

https://github.com/rust-lang/rust-analyzer/blob/2b866390189d99191c51e1fe154e94a933c24b0f/crates/hir-ty/src/display.rs#L1484-L1518

and this moves lifetimes before the type and const parameters as it assumes that the input it gets has types and const parameters before the lifetime parameters, as `generics.iter()` gives them in this order;

https://github.com/rust-lang/rust-analyzer/blob/2b866390189d99191c51e1fe154e94a933c24b0f/crates/hir-ty/src/generics.rs#L88-L95

But this function does not do such re-ordering to parent parameters and we only do such re-ordering manually in the block I've modified in this PR.

So, in other types the parent generics can be renderered in the rear position of the generic parameters and I'm not sure whether that's correct.

And If we should do such re-ordering everywhere we have to thinkabout grand-parent generics as well, I think 🤔 

Furthermore, in `fn generic_args_sans_defaults`, we assume that the default parameters are comming in the consecutive last some parameters and this is correct in general in proper rust syntax(L1476)

https://github.com/rust-lang/rust-analyzer/blob/2b866390189d99191c51e1fe154e94a933c24b0f/crates/hir-ty/src/display.rs#L1431-L1482 

But I think that this can be problematic if we have both non-empty self parameters and parent parameters, and they are have some both defaults and non defaults.
In this case, aren't we losing some non-default generic parameter of the parent? 🤔

Reply:
So, I'm just trying to specify the exact problems we have here with some experiments 😅

Reply:
That might be leftovers, we used to have lifetimes after type/const params, but we've since fixed that order.

> grand-parent generics as well, I think 

I don't think Rust currently has that anywhere?

Reply:
In general the order ought to be https://github.com/rust-lang/rust-analyzer/blob/bd4785a6f017ec52bc0fb6c061c5f42228946462/crates/hir-ty/src/generics.rs#L1-L9
now